### PR TITLE
Embeds - Formula

### PR DESCRIPTION
**Pull request by garv-shah**
Linked to issue here: https://github.com/visual-space/visual-editor/issues/108
Added preliminary maths support with quite a few bugs, needs refining

### CODE REVIEW

**Comment by adrian-moisa** [comment]
I had a tough time lately finding time. There's a lot to do in the backlog and the dayjob was rather demanding lately. I'll review carefully these days the commit to see what can be improved. Expect more feedback.

File: lib/embeds/services/image.utils.dart [hunk 1]
```diff
@@ -50,18 +51,61 @@ Image imageByUrl(String imageUrl,
...
+    url = url.replaceAll(
+        'https://drive.google.com/file/d/', 'https://drive.google.com/uc?id=');
+  }
+
+  // Changes Discord domain
+  if (url.contains('cdn.discordapp.com')) {
+    url = url.replaceAll('cdn.discordapp.com', 'media.discordapp.net');
+  }
+
+  // Uses CORS proxy for web
```
Comment (adrian-moisa) [inline lib/embeds/services/image.utils.dart:103]: I don't believe this should be part of the library. There should be a configuration feature. And a nice thought trough, API for it.

File: lib/embeds/widgets/default-embed-builder.dart [hunk 2]
```diff
@@ -1,7 +1,9 @@
 import 'package:flutter/cupertino.dart';
 import 'package:flutter/foundation.dart';
 import 'package:flutter/material.dart';
+import 'package:flutter_math_fork/flutter_math.dart';
```
Comment (adrian-moisa) [inline lib/embeds/widgets/default-embed-builder.dart:4]: I did some research and found that flutter math is no longer maintained. 

https://pub.dev/packages/flutter_math_fork
This repo claims so

There are a few options around. Any strong reason to chose flutter_math?

https://fluttergems.dev/latex/

File: lib/embeds/widgets/default-embed-builder.dart [hunk 3]
```diff
@@ -199,6 +200,13 @@ Widget defaultEmbedBuilder(
         readOnly: readOnly,
       );
 
+    case BlockEmbedM.formulaType:
+      final formula = node.value.data;
+      final style = node.style.attributes['style'];
+      print(node.style);
```
Comment (adrian-moisa) [inline lib/embeds/widgets/default-embed-builder.dart:206]: You should not have print statements in production

File: lib/embeds/widgets/default-embed-builder.dart [hunk 4]
```diff
@@ -199,6 +200,13 @@ Widget defaultEmbedBuilder(
...
+      return Math.tex(formula, textStyle: Theme.of(context).textTheme.subtitle1);
```
Comment (adrian-moisa) [inline lib/embeds/widgets/default-embed-builder.dart:207]: I guess we should not hardcode to one single style. At least the font size could be taken in account

File: lib/embeds/widgets/youtube-video-app.dart [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import 'package:flutter/gestures.dart';
 import 'package:flutter/material.dart';
 import 'package:url_launcher/url_launcher.dart';
-import 'package:youtube_player_flutter/youtube_player_flutter.dart';
+import 'package:youtube_player_flutter_quill/youtube_player_flutter_quill.dart';
```
Comment (adrian-moisa) [inline lib/embeds/widgets/youtube-video-app.dart:4]: I'm struggling to see where's the source code for this repository. I don't trust the library. Can you explain more what's the story?

**Comment by adrian-moisa** [comment]
Here's some feedback regarding the current PR. Won't merge it for now. We need to improve the support and code quality.

- Provide a formula demo page with multiple examples
- Formula size is not sensitive to line size
- Undo from keyboard does not restore the formula. OBJ is restored.
- When tapping the formula we should be able to edit it. Now you have to delete and start over.
- Setup demo page for image
- Setup demo page for YT player
- I don't believe this should be part of the library. There should be a configuration feature. And a nice thought trough, API for it.

```dart
  if (kIsWeb && !url.contains('firebasestorage')) {
    return 'https://corsproxy.garvshah.workers.dev/?$url';
  } else {

```

- I did some research and found that flutter math is no longer maintained.
This repo claims so: https://pub.dev/packages/flutter_math_fork
There are a few options around. Any strong reason to chose flutter_math? https://fluttergems.dev/latex/

- You should not have print statements in production
- I'm struggling to see where's the source code for this repository. I don't trust the library. Can you explain more what's the story?
- We need to setup some testing to demo that it works.

Sorry for the late review, I had an extreme amount of backlog to catch up with.

**Comment by adrian-moisa** [comment]
Closed for now since the concerns have not been resolved. We will add this feature sooner or later. It's already planned. We will use this PR as a guide on how to continue and improve.

**Comment by garv-shah** [comment]
All good, if I ever find the time I'll be sure to submit a better PR addressing the concerns, thanks!

**Pull request closed without merging**